File: 2017/day05/day05.cpp
#include "../String.h"
#include "../Task.h"

class Day05 : public AbstractTask {
 public:
  std::string part_one() override {
    auto arr = parse_input(input());

    size_t curr = 0;
    size_t steps = 0;
    while (true) {
      steps++;
      int jumps = arr[curr];
      size_t next = curr + jumps;

      arr[curr]++;

      if (next >= arr.size()) {
        break;
      }

      curr = next;
    }

    return std::to_string(steps);
  }

  std::string part_two() override {
    auto arr = parse_input(input());

    size_t curr = 0;
    size_t steps = 0;
    while (true) {
      steps++;
      int offset = arr[curr];
      size_t next = curr + offset;

      if (offset >= 3) {
        arr[curr]--;
      } else {
        arr[curr]++;
      }

      if (next >= arr.size()) {
        break;
      }

      curr = next;
    }

    return std::to_string(steps);
  }

 private:
  std::string input() { return read_input("inputs/day05_input.txt"); }

  static std::vector<int> parse_input(const std::string& input) {
    std::vector<std::string> lines = split(input, '\n');

    std::vector<int> vec;
    std::transform(
        lines.begin(), lines.end(), std::back_inserter(vec),
        [](const std::string& line) -> int { return std::stoi(line); });

    return vec;
  }
};

File: 2021/day05/day05.rb
# frozen_string_literal: true

require_relative "../task"

class Day05
  include Task

  def part_one
    lines = parse_input(read_input)
    lines = lines.filter { |item| item.vertical? || item.horizontal? }

    map = {}
    lines.each do |line|
      if line.vertical?
        if line.start_point.y < line.end_point.y
          top = line.start_point.y
          bottom = line.end_point.y
        else
          top = line.end_point.y
          bottom = line.start_point.y
        end

        (top..bottom).each do |y|
          key = "#{line.start_point.x},#{y}"
          map[key] = 0 unless map.key?(key)

          map[key] = map[key] + 1
        end

        next
      end

      if line.horizontal?
        if line.start_point.x < line.end_point.x
          left = line.start_point.x
          right = line.end_point.x
        else
          left = line.end_point.x
          right = line.start_point.x
        end

        (left..right).each do |x|
          key = "#{x},#{line.start_point.y}"
          map[key] = 0 unless map.key?(key)

          map[key] = map[key] + 1
        end

        next
      end

      raise StandardError, "Diagonal"
    end

    map.select { |k, v| v > 1 }.size.to_s
  end

  def part_two
    "-"
  end

  private

  def parse_input(input)
    input.split("\n").map do |item|
      raw_points = item.split(" -> ")

      start_point = Point.from_str(raw_points[0])
      end_point = Point.from_str(raw_points[1])

      Line.new(start_point, end_point)
    end
  end
end

class Point
  attr_reader :x, :y

  def initialize(x, y)
    @x = x
    @y = y
  end

  def self.from_str(str)
    tokens = str.split(",")
    self.new(tokens[0].to_i, tokens[1].to_i)
  end
end

class Line
  attr_reader :start_point, :end_point

  def initialize(start_point, end_point)
    @start_point = start_point
    @end_point = end_point
  end

  def horizontal?
    start_point.y == end_point.y
  end

  def vertical?
    start_point.x == end_point.x
  end
end